Source/Scene/ModelExperimental/GeometryPipelineStage.js

```diff
@@ -43,145 +92,224 @@ GeometryPipelineStage.process = function (renderResources, primitive) {
   }
 
   shaderBuilder.addVertexLines([GeometryStageVS]);
-  shaderBuilder.addVarying("vec3", "v_positionEC");
+  shaderBuilder.addFragmentLines([GeometryStageFS]);
 };
 
-function processAttribute(
-  renderResources,
-  attribute,
-  attributeIndex,
-  setIndexedAttributeInitializationLines,
-  customAttributeInitializationLines
-) {
+function processAttribute(renderResources, attribute, attributeIndex) {
+  var shaderBuilder = renderResources.shaderBuilder;
+  var attributeInfo = ModelExperimentalUtility.getAttributeInfo(attribute);
+
+  addAttributeToAttributesArray(
+    renderResources.attributes,
+    attribute,
+    attributeIndex
+  );
+  addAttributeDeclaration(shaderBuilder, attributeInfo);
+  addVaryingDeclaration(shaderBuilder, attributeInfo);
+
+  // For common attributes like positions, normals and tangents, the code is
+  // already in GeometryStageVS, we just need to enable it
+  if (defined(attribute.semantic)) {
+    addSemanticDefine(shaderBuilder, attribute);
+  }
+
+  // Some GLSL code must be dynamically generated
+  updateAttributesStruct(shaderBuilder, attributeInfo);
+  updateInitialzeAttributesFunction(shaderBuilder, attributeInfo);
+  updateSetDynamicVaryingsFunction(shaderBuilder, attributeInfo);
+}
+
+function addSemanticDefine(shaderBuilder, attribute) {
   var semantic = attribute.semantic;
   var setIndex = attribute.setIndex;
-  var attributeType = attribute.type;
-
-  var shaderBuilder = renderResources.shaderBuilder;
+  switch (semantic) {
+    case VertexAttributeSemantic.NORMAL:
+      shaderBuilder.addDefine("HAS_NORMALS");
+      break;
+    case VertexAttributeSemantic.TANGENT:
+      shaderBuilder.addDefine("HAS_TANGENTS");
+      break;
+    case VertexAttributeSemantic.FEATURE_ID:
+    case VertexAttributeSemantic.TEXCOORD:
+    case VertexAttributeSemantic.COLOR:
+      shaderBuilder.addDefine("HAS_" + semantic + "_" + setIndex);
+  }
+}
 
-  var variableName;
-  var varyingName;
-  var glslType = AttributeType.getGlslType(attributeType);
-
-  if (defined(semantic)) {
-    variableName = VertexAttributeSemantic.getVariableName(semantic, setIndex);
-    varyingName = "v_" + variableName;
-
-    switch (semantic) {
-      case VertexAttributeSemantic.NORMAL:
-        shaderBuilder.addDefine("HAS_NORMALS");
-        break;
-      case VertexAttributeSemantic.TANGENT:
-        shaderBuilder.addDefine("HAS_TANGENTS");
-        break;
-      case VertexAttributeSemantic.FEATURE_ID:
-      case VertexAttributeSemantic.TEXCOORD:
-      case VertexAttributeSemantic.COLOR:
-        shaderBuilder.addDefine("HAS_" + semantic + "_" + setIndex);
-        setIndexedAttributeInitializationLines.push(
-          "    #ifdef HAS_" + semantic + "_" + setIndex
-        );
-        setIndexedAttributeInitializationLines.push(
-          "    " + varyingName + " = a_" + variableName + ";"
-        );
-        setIndexedAttributeInitializationLines.push("    #endif");
-    }
+function addAttributeToAttributesArray(
```

Comment:
Let's find a better name for this function. I think it'd be cleaner to just add this pass in the `renderResources` and call this function `addAttributeToRenderResources`